docs/spec.md

```diff
@@ -5918,3 +5889,142 @@ continue execution without raising the corresponding status flag; similar to
+`[dim(lhs, dim1) for dim1 in lhs_batching_dimensions] ==
+[dim(rhs, dim2) for dim2 in rhs_batching_dimensions]`.
+
+Thanks to ellipses, it is often possible to avoid working at the level of
+individual scalars. However, in some tricky cases, lower-level semi-informal
+syntax may be used like in the `start_indices[bi0, ..., :, ..., biN]` formula
+from the `gather` specification. In the service of conciseness, we don't
+provide an exact formalism for translating such syntax to vanilla Python, in
+hopes that it is still intuitively understandable on case-by-case basis.
+Please let us know if some specific formulas look opaque, and we'll try to
+improve them.
+
+C) The final noteworthy notational vehicle that we employ is implicit
+broadcasting. While the StableHLO opset doesn't support implicit broadcasting,
+the formulas do, also in the service of conciseness. In a nutshell, if a scalar
+is used in a context where a tensor is expected, the scalar is broadcasted to
+the expected shape.
+
+To continue the `dot_general` example, here's another constraint:
+`0 <= lhs_batching_dimensions < rank(lhs)`. As defined in the `dot_general`
+specification, `lhs_batching_dimensions` is a tensor, however both `0` and
+`rank(lhs)` are scalars. After we apply implicit broadcasting, the formula will
+become `[0, ..., 0] <= lhs_batching_dimensions < [rank(lhs), ..., rank(lhs)]`.
+
+### Names
+
+In formulas, lexical scope includes: 1) global functions, 2) member definitions,
+3) local definitions. The list of global functions is provided below. The list
+of element definitions depends on the program element that the notation is
+applied to:
+
+* For operations, member definitions include names introduced in "Inputs" and
+  "Outputs" sections.
+* For everything else, member definitions include structural parts of the
+  program element, named after the corresponding EBNF non-terminals. Most of
+  the time, the names of these structural parts are obtained by converting the
+  names of the non-terminals to snake case (e.g. `IntegerLiteral` =>
+  `integer_literal`), but sometimes names get abbreviated in the process (e.g.
+  `QuantizationStorageType` => `storage_type`) in which case the names are
+  introduced explicitly similarly to "Inputs" / "Outputs" sections in operation
+  specifications.
+* Additionally, member definitions always include `self` to refer to the
+  corresponding program element.
+
+### Values
+
+TODO: Describe the types of values that are involved in evaluating formulas:
+1\) Value (actual values e.g. constants), 2) Placeholder (future values e.g.
+inputs / outputs from the perspective of constraints), 3) Type.
+
+### Functions
+
+TODO: Describe the functions available in formulas.
+
+// 1) Deconstruction of types
+// foo_bar(any: Value | Placeholder | Type) -> Type
+// for every member definition foo_bar in the respective type.
+
+// 2) Construction of types
+// No functions, just use notation
+
+// 3) Functions on types
```

Comment:
We may add here the functions like
`is_token`, `is_complex`?

Reply:
Right. This is covered by "is_foo(any: Value | Placeholder | Type) -> Value".